### PR TITLE
Best practices page

### PR DESCRIPTION
**Pull request by vladaionescu**
Preview: https://app.gitbook.com/o/-MB5s2W0v8S4xFU5ELti/s/iZfkHzyR9lPa2mtTFR0D/best-practices

### CODE REVIEW

File: docs/best-practices/best-practices.md [hunk 1]
```diff
@@ -0,0 +1,370 @@
+# Best practices
+
+Although Earthly has been designed to be unambiguous about what command to use for the job, writing Earthfiles can sometimes still be tricky, when it comes to nuances. As you try to accomplish certain tasks, you may find that sometimes the same result can be achieved using more than one technique. Or so it might seem.
+
+Below we list some of the best practices that we have found to be useful in writing Earthfiles, with a focus on certain commands or techniques that seem similar, but aren't really, but also on some key points that we have seen newcomers stumble into.
+
+## `COPY` only the minimal amount of files. Avoid copying `.git`
+
+A typical mistake is to `COPY` entire large directories into the build environment and only using a subset of the files within them. Or worse, copying the entire repository (which might also include `.git`) for no good reason.
```
Comment (brandonSc) [inline docs/best-practices/best-practices.md:9]: With Docker you might have a `.dockerignore` to exclude something like `.git`. Do we support something similar?

**Comment by brandonSc** [comment]
Well written and super useful. I learned quite a few things from this guide!

File: docs/best-practices/best-practices.md [hunk 2]
```diff
@@ -0,0 +1,1015 @@
...
+
+Note, however, that none of these cases are needed in a CI environment, and ultimately these commands are not regularly tested by a CI, which means they may break more frequently.
+
+## Pattern: Optionally `LOCALLY`
+
+In certain cases, it may be desirable to execute certain targets on the host machine, rather than in the sandboxed build environment, for debugging purposes. However, we need most of the target to execute in strict mode in CI. The solution to this is to use a target that can be optionally executed via `LOCALLY`. Here is an example:
+
+Suppose we wanted the following target to be executed on against the host's Docker daemon:
+
+```Dockerfile
```
Comment (ingwarsw) [inline docs/best-practices/best-practices.md:144]: Use `Earthfile` syntax would work here?

File: docs/best-practices/best-practices.md [hunk 3]
```diff
@@ -0,0 +1,1015 @@
...
+        --service db \
+        --load=+integration-test
+    RUN docker-compose up integration
+END
+```
+
+We could have an equivalent `LOCALLY` target:
+
+```Dockerfile
+LOCALLY
```
Comment (ingwarsw) [inline docs/best-practices/best-practices.md:158]: Use `Earthfile` syntax would work here?

File: docs/best-practices/best-practices.md [hunk 4]
```diff
@@ -0,0 +1,1015 @@
...
+WITH DOCKER --compose docker-compose.yml \
+        --service db \
+        --load=+integration-test
+    RUN docker-compose up integration
+END
+```
+
+We could have an equivalent `LOCALLY` target:
+
+```Dockerfile
```
Comment (ingwarsw) [inline docs/best-practices/best-practices.md:157]: Use `Earthfile` syntax would work here?

File: docs/guides/udc.md [hunk 5]
```diff
@@ -72,3 +72,28 @@ PRINT_VAR:
     COMMAND
     RUN echo "$a_global_var"
 ```
+
+## Targets vs UDCs
+
+Earthly targets and UDCs are Earthly's core primitives for organizing build recipes. They encapsulate build logic, and from afar they look pretty similar. However, the use-cases for each are vastly different.
+
+In general, targets are used to produce specific build results, while UDCs are used as a way to reuse build logic, when certain commands are repeated in multiple places. As a real-world analogy, targets are more like factories, while UDCs are more like components that are used to put together factories.
```
Comment (adamgordonbell) [inline docs/guides/udc.md:80]: What about adding "UDCs work like function or methods call in an imperative programming language (c, javascript, python). Much like function calls it's helpful to imagine UDCs being executed by being inlined into the call site but in a separate variable scope"

File: docs/best-practices/best-practices.md [hunk 6]
```diff
@@ -0,0 +1,1015 @@
...
+
+Examples of valid cases where `LOCALLY` may be used are:
+
+* installing dependencies on the host machine (e.g. to help IDEs provide better suggestions)
+* executing tests on the host docker daemon, to help with inspection and debugging
+* executing development commands which would otherwise require copying very large amounts of files to the sandboxed build environment
+
+Note, however, that none of these cases are needed in a CI environment, and ultimately these commands are not regularly tested by a CI, which means they may break more frequently.
+
+## Pattern: Optionally `LOCALLY`
```
Comment (adamgordonbell) [inline docs/best-practices/best-practices.md:138]: Good tip!

File: docs/best-practices/best-practices.md [hunk 7]
```diff
@@ -0,0 +1,1015 @@
...
+integration-test:
+    FROM some-other-image:latest
+    DO github.com/earthly/lib+INSTALL_DIND
+    COPY docker-compose.yml ./
+    WITH DOCKER --compose docker-compose.yml
+        RUN ...
+    END
+```
+
+A simple way to fix this is to use an earthly-provided [UDC](../guides/udc.md) to install Docker engine before the `COPY` command.
```
Comment (adamgordonbell) [inline docs/best-practices/best-practices.md:982]: Worth mentioning that the UDC is fastest on alpine?

**Review by adamgordonbell — APPROVED** [review]
Great job. This will be helpful for sure.

**Review by brandonSc — APPROVED** [review]

File: docs/guides/udc.md (hunk 5 above)
Comment (vladaionescu) [reply docs/guides/udc.md]: Good idea

**Pull request merged**